### PR TITLE
add support for iterators

### LINKED ISSUE #198: Request Body as Iterable/Iterator
Current flow is to just use `Readable.from(i)`. Which is fine. But we could make this faster.

### PR DESCRIPTION
**Pull request by Ethan-Arrowood**
closes #198

### CODE REVIEW

File: lib/core/util.js [hunk 1]
```diff
@@ -10,6 +10,11 @@ function isStream (body) {
   return !!(body && typeof body.on === 'function')
 }
 
+function isGenerator (fn) {
+  const name = fn.constructor.name
```
Comment (ronag) [inline lib/core/util.js:14]: fn could be null and/or not a function

File: lib/core/request.js [hunk 2]
```diff
@@ -74,6 +75,8 @@ class Request {
         // TODO: Ignore error if body has ended?
         this.onError(err)
       })
+    } else if (util.isGenerator(body)) {
+      this.body = Readable.from(body())
     } else if (util.isBuffer(body)) {
```
Comment (ronag) [inline lib/core/request.js:80]: We could also add support for iterable at the same time?

**Review by ronag — COMMENTED** [review]
Think out load.

- I wonder if whether it is iterable and/or generator function we should support.
- Should we assume a generator function is retryable?

File: lib/core/request.js (hunk 2 above)
Comment (Ethan-Arrowood) [reply lib/core/request.js:80]: Ahh so checking for `[Symbol.iterator]` or something like that right?

**Comment by Ethan-Arrowood** [comment]
I think if I make this work for just iterators in general then it'll automatically support generators. I'll give that a try now

File: lib/core/request.js (hunk 2 above)
Comment (ronag) [reply lib/core/request.js:80]: check what Readable.from does

Comment (Ethan-Arrowood) [reply lib/core/request.js]: yep so in that case if its an iterable object then we can pass it directly to Readable.from

if its a generator function we have to pass the return value of it to Readable.from

**Comment by Ethan-Arrowood** [comment]
So I added the generator support in as a separate commit. easy to revert if we don't want that

**Comment by Ethan-Arrowood** [comment]
Strange, not sure why tests failed this time. node 14 works on my mac

**Comment by ronag** [comment]
> Strange, not sure why tests failed this time. node 14 works on my mac

Looks like a flaky test.

File: lib/core/util.js [hunk 3]
```diff
@@ -10,6 +10,15 @@ function isStream (body) {
   return !!(body && typeof body.on === 'function')
 }
 
+function isIterator (fn) {
+  return !!(fn != null && typeof fn[Symbol.iterator] === 'function')
```
Comment (ronag) [inline lib/core/util.js:14]: asyncIterator

Comment (ronag) [reply lib/core/util.js:14]: `isIterable`

Comment (ronag) [reply lib/core/util.js:14]: `isIterator` is a different check, i.e. `typeof next === 'function'`.

File: lib/core/request.js [hunk 4]
```diff
@@ -78,6 +79,10 @@ class Request {
       this.body = body.length ? body : null
     } else if (typeof body === 'string') {
       this.body = body.length ? Buffer.from(body) : null
+    } else if (util.isIterator(body)) { // check this after buffer and string so it doesn't return false positives
+      this.body = Readable.from(body)
+    } else if (util.isGenerator(body)) {
+      this.body = Readable.from(body())
```
Comment (ronag) [inline lib/core/request.js:85]: I think you might be able to skip the checks here and just pass it to `Readable.from` and catch INVALID_ARG_TYPE errors from there. That way you don't need to re-implement the type checks.

Comment (Ethan-Arrowood) [reply lib/core/request.js:85]: Hmm okay I also found this line in the [documentaion](https://nodejs.org/dist/latest-v14.x/docs/api/stream.html#stream_stream_readable_from_iterable_options):

> Calling Readable.from(string) or Readable.from(buffer) will not have the strings or buffers be iterated to match the other streams semantics for performance reasons.

So like is it potentially safe to just always pass anything but an existing stream to Readable.from?

**Review by mcollina — CHANGES_REQUESTED** [review]
I might be too late to the party but I do not think that using `Readable.from()` is a good way of doing this. It's the simplest way... however it would add significant overhead. If we want those to be fast we should not wrap the generator in a string.

**Comment by Ethan-Arrowood** [comment]
Should I use a `for...of` loop instead?

**Comment by mcollina** [comment]
I would use `pipeline` somewhere down, it handles those (or copy what they are doing).

**Comment by Ethan-Arrowood** [comment]
And would `this.body` be the writestream in that case?

**Comment by Ethan-Arrowood** [comment]
If the new try catch method is better I'll delete the util methods

**Comment by ronag** [comment]
You basically need to replicate all the logic behind útil.isStream but with útil.isIterable

**Comment by Ethan-Arrowood** [comment]
what is missing from my current implementation of `util.isIterable`? 

Also, with the try/catch do we even need those util methods?

**Comment by ronag** [comment]
> what is missing from my current implementation of util.isIterable?

Nothing is missing.

> Also, with the try/catch do we even need those util methods?

No really...

But as @mcollina pointed out, this implementation probably works but is not as fast as it could be.

File: lib/core/request.js [hunk 5]
```diff
@@ -79,7 +80,15 @@ class Request {
     } else if (typeof body === 'string') {
       this.body = body.length ? Buffer.from(body) : null
     } else {
-      throw new InvalidArgumentError('body must be a string, a Buffer or a Readable stream')
+      try {
+        this.body = Readable.from(typeof body === 'function' ? body() : body)
```
Comment (ronag) [inline lib/core/request.js:84]: I think we might want to skip generator support and only to iterable. Generators have some semantics we might want to give more consideration to.

Comment (Ethan-Arrowood) [reply lib/core/request.js:84]: Okay I'll remove generator support

**Comment by Ethan-Arrowood** [comment]
Okay sounds good. I'll remove the utils and rebase for the conflict. @mcollina if you have any more pointers on making this faster let me know; i'd like to look into it and give it a shot.

**Comment by ronag** [comment]
> if you have any more pointers on making this faster let me know; i'd like to look into it and give it a shot.

In order to make this faster you should avoid converting it to stream and iterate it directly in the `write(client, request)` method. Also take into account all the places where there is logic based on `util.isStream(body)` and implement accordingly.

**Comment by Ethan-Arrowood** [comment]
> In order to make this faster you should avoid converting it to stream and iterate it directly in the write(client, request) method. Also take into account all the places where there is logic based on util.isStream(body) and implement accordingly.

Ahh I understand better now. Okay I'll give it a shot too

**Comment by Ethan-Arrowood** [comment]
Help me think through something please:

The iterable object a user passes to the body, does it need to always yield a string or buffer? If so do I need to check for that when I'm iterating over the iterable inside of `write` and handle accordingly?

**Comment by ronag** [comment]
I think you should check for string or buffer and throw if something else

**Comment by Ethan-Arrowood** [comment]
And I should ditch generator support?

File: lib/core/client.js [hunk 6]
```diff
@@ -1153,6 +1153,36 @@ function write (client, request) {
...
+    let chunks = stringOrBuffer(result.value) ? result.value : [result.value]
+    while (true) {
+      if (result.done) break
+      result = iterator.next()
+      if (stringOrBuffer(result.value)) {
+        chunks += result.value
+      } else {
+        chunks.push(result.value)
+      }
+    }
```
Comment (ronag) [inline lib/core/client.js:1177]: Close, I would probably do it like this:

```js
async function write(iterable, socket) {
  try {
    if (!contentLength) {
      // chunked transfer
    }
    for await (const chunk of iterable) {
      socket.write(chunk)
    }
    socket.write('\r\n', 'ascii')

    request.body = null
  } catch (err) {
    util.destroy(socket, err)
  } 
}
```

The iterable could be huge and not appropriate to read into memory.

**Review by ronag — CHANGES_REQUESTED** [review]

File: lib/core/client.js (hunk 6 above)
Comment (ronag) [reply lib/core/client.js:1177]: You basically need to do the same thing as for streams.

Comment (Ethan-Arrowood) [reply lib/core/client.js:1177]: Ahh okay I'll try this out. Thank you!

Comment (Ethan-Arrowood) [reply lib/core/client.js:1177]: making some progress here, but since this is taking me quite a bit of time i don't want to impede the overall progress on the library. @ronag would you be interested in maybe doing a pair programming call and working through this together? No worries if you're too busy right now; i'll keep plugging away at it

Comment (ronag) [reply lib/core/client.js:1177]: This week is swamped for me. If you are still working on it next week we could try to do some pair programming.

Comment (Ethan-Arrowood) [reply lib/core/client.js:1177]: Okay sounds good!

**Review by mcollina — APPROVED** [review]
lgtm

File: lib/core/client.js [hunk 7]
```diff
@@ -1153,6 +1153,29 @@ function write (client, request) {
...
+    try {
+      if (contentLength === null) {
+        socket.write(`${header}transfer-encoding: chunked\r\n`, 'ascii')
+      } else {
+        socket.write(`${header}content-length: ${contentLength}\r\n\r\n`, 'ascii')
+      }
+      for await (const chunk of body) {
+        if (typeof chunk !== 'string' && !util.isBuffer(chunk)) {
+          throw new Error('Body iterable must yield either string or buffer')
+        } else if (socket.write(chunk)) {
```
Comment (ronag) [inline lib/core/client.js:1166]: Chunked encoding needs to write length becore each chunk

**Comment by Ethan-Arrowood** [comment]
From [http watch](https://www.httpwatch.com/httpgallery/chunked/)

> However, with chunked encoding you could just write the data one row at a time and write a zero sized chunk when the end of the query was reached.

I think I'm missing the zero sized chunk at the end. Will try and add this tomorrow

**Comment by Ethan-Arrowood** [comment]
So by making `write` async, i've now broken other things. Hmmmm

**Comment by Ethan-Arrowood** [comment]
Hmm removing async seems to solve things but I guess we lose support for asyncIterators? Or maybe not? This is really pressing my understanding of async things

File: lib/core/util.js [hunk 8]
```diff
@@ -10,12 +10,18 @@ function isStream (body) {
   return !!(body && typeof body.on === 'function')
 }
 
+function isIterable (obj) {
+  return !!(obj != null && (typeof obj[Symbol.iterator] === 'function' || typeof obj[Symbol.asyncIterator] === 'function'))
```
Comment (Ethan-Arrowood) [inline lib/core/util.js:14]:
```suggestion
  return !!(obj != null && typeof obj[Symbol.iterator] === 'function')
```
Without async support in `write` method I don't think `asyncIterators` will work any longer

Comment (ronag) [reply lib/core/util.js:14]: async iterators should work

File: lib/core/client.js [hunk 9]
```diff
@@ -1153,6 +1153,41 @@ function write (client, request) {
...
       .on('error', onFinished)
       .on('close', onFinished)
+  } else if (util.isIterable(body)) {
+    try {
+      if (contentLength === null) {
+        socket.write(`${header}transfer-encoding: chunked\r\n`, 'ascii')
+      } else {
+        socket.write(`${header}content-length: ${contentLength}\r\n\r\n`, 'ascii')
+      }
+      for (const chunk of body) {
```
Comment (ronag) [inline lib/core/client.js:1163]: for await

Comment (ronag) [reply lib/core/client.js:1163]: move this part out to a `async function writeIterable` and maybe refactor the above into a `function writeStream`

Comment (Ethan-Arrowood) [reply lib/core/client.js:1163]: when I make the write method async it breaks a bunch of stuff since write is called synchronously in a while loop. any ideas to change that?

Comment (ronag) [reply lib/core/client.js:1163]: Yes... what I wrote above 😊

**Review by ronag — CHANGES_REQUESTED** [review]
You need to add a check here as well:

```js
    if (client.running && (util.isStream(request.body) || util.isIterable(request.body)) {
      // Request with stream body can error while other requests
      // are inflight and indirectly error those as well.
      // Ensure this doesn't happen by waiting for inflight
      // to complete before dispatching.

      // Request with stream body cannot be retried.
      // Ensure that no other requests are inflight and
      // could cause failure.
      return
    }
```

Also a test that fails without it.

Also please try to make sure `npm run coverage` is at `100`.

File: lib/core/client.js [hunk 10]
```diff
@@ -916,6 +916,42 @@ function _resume (client) {
...
+  try {
+    if (contentLength === null) {
+      socket.write(`${header}transfer-encoding: chunked\r\n`, 'ascii')
+    } else {
+      socket.write(`${header}content-length: ${contentLength}\r\n\r\n`, 'ascii')
+    }
+    for await (const chunk of body) {
+      if (typeof chunk !== 'string' && !util.sBuffer(chunk)) {
+        throw new Error('Body iterable must yield either string or buffer')
+      }
```
Comment (ronag) [inline lib/core/client.js:929]: I'm not sure you need this check? socket.write() will fail anyway

File: lib/core/client.js [hunk 11]
```diff
@@ -916,6 +916,42 @@ function _resume (client) {
...
+      if (socket.write(chunk)) {
+        continue
+      } else if (socket.destroyed) {
+        return
+      } else {
+        await EventEmitter.once(socket, 'drain')
+      }
+    }
+    socket.write('\r\n0\r\n\r\n', 'ascii')
+    socket.cork()
```
Comment (ronag) [inline lib/core/client.js:949]: this cork looks wrong?

**Comment by Ethan-Arrowood** [comment]
Just working on unit tests now! Some branches i'm having a hard time reaching but I'm getting there 👍

File: lib/core/client.js [hunk 12]
```diff
@@ -909,7 +909,10 @@ function _resume (client, sync) {
       request.body = null
     }
 
-    if (client.running && util.isStream(request.body)) {
+    if (client.running &&
```
Comment (Ethan-Arrowood) [inline lib/core/client.js:912]: @ronag how do I get the request into a "running" state while also still having the request.body defined?

Comment (ronag) [reply lib/core/client.js:912]: Create two request with a non empty stream body?

Comment (Ethan-Arrowood) [reply lib/core/client.js:912]: haven't forgotten about this. I tried again but didn't make much progress. I'm going to try another way later by manually stepping through the execution and seeing what I will need to do to get the code down a certain path

Comment (Ethan-Arrowood) [reply lib/core/client.js:912]: I think I gotta ask for help on this. Really struggling to write tests to get to some of the edge cases in the writeIterable method. @ronag would you like to pair on this? otherwise, if you want to author the necessary tests that is okay with me

Comment (ronag) [reply lib/core/client.js:912]: I'll get back to you.

Comment (ronag) [reply lib/core/client.js:912]: @Ethan-Arrowood We already have tests which hit this when using streams? Can't you just find the test and replace the stream part with an iterable?

File: lib/core/client.js [hunk 13]
```diff
@@ -909,7 +909,10 @@ function _resume (client, sync) {
...
+      !util.isBuffer(request.body) &&
+      typeof request.body !== 'string' &&
+      (util.isStream(request.body) || util.isIterable(request.body))) {
```
Comment (ronag) [inline lib/core/client.js:915]: This looks wrong, it should just be `if (client.running && (isStream(request.body) || isIterable(request.body)))`

File: lib/core/client.js [hunk 14]
```diff
@@ -929,6 +932,41 @@ function _resume (client, sync) {
...
+  try {
+    if (contentLength === null) {
+      socket.write(`${header}transfer-encoding: chunked\r\n`, 'ascii')
+    } else {
+      socket.write(`${header}content-length: ${contentLength}\r\n\r\n`, 'ascii')
+    }
+    for await (const chunk of body) {
+      // if (typeof chunk !== 'string' && !util.sBuffer(chunk)) {
+      //   throw new Error('Body iterable must yield either string or buffer')
+      // }
```
Comment (ronag) [inline lib/core/client.js:945]: Uncommented?

File: lib/core/request.js [hunk 15]
```diff
@@ -49,8 +49,10 @@ class Request {
       this.body = body.length ? body : null
     } else if (typeof body === 'string') {
       this.body = body.length ? Buffer.from(body) : null
+    } else if (util.isIterable(body)) {
+      this.body = body
     } else {
-      throw new InvalidArgumentError('body must be a string, a Buffer or a Readable stream')
+      throw new InvalidArgumentError('body must be a string, a Buffer, a Readable stream, or an iterable')
```
Comment (ronag) [inline lib/core/request.js:55]: `Iterable, AsyncIterable`

File: lib/core/client.js (hunk 12 above)
Comment (Ethan-Arrowood) [reply lib/core/client.js:912]: yeah thats what I thought i must've introduce some regressions cause the coverage reporter was saying some of these lines were not getting tested.

File: lib/core/client.js [hunk 16]
```diff
@@ -929,6 +932,41 @@ function _resume (client, sync) {
...
+        return
+      } else {
+        await EventEmitter.once(socket, 'drain')
+      }
+    }
+    socket.write('\r\n0\r\n\r\n', 'ascii')
+  } catch (err) {
+    util.destroy(socket, err)
+  }
+}
```
Comment (ronag) [inline lib/core/client.js:968]: Take  another look at the streams version. You need to e.g. set `client[kWriting]`.

Comment (ronag) [reply lib/core/client.js:968]: Missing checks with `bytesWritten`

Comment (ronag) [reply lib/core/client.js:968]: You basically want to almost the exact same thing but with a for await loop instead of stream events.

File: lib/core/client.js (hunk 14 above)
Comment (Ethan-Arrowood) [reply lib/core/client.js:945]: not sure why i commented this; will fix.

File: lib/core/client.js [hunk 17]
```diff
@@ -909,7 +909,10 @@ function _resume (client, sync) {
       request.body = null
     }
 
-    if (client.running && util.isStream(request.body)) {
+    if (client.running &&
+      !util.isBuffer(request.body) &&
+      typeof request.body !== 'string' &&
```
Comment (ronag) [inline lib/core/client.js:914]: I don't think body can be a string here?

Comment (Ethan-Arrowood) [reply lib/core/client.js:914]: It failed without these checks. I'll have to look back on it though

File: lib/core/client.js (hunk 12 above)
Comment (ronag) [reply lib/core/client.js:912]: This needs significantly more tests. Preferably we should just port most of the body as stream test into a body as iterable test.

File: lib/core/client.js (hunk 17 above)
Comment (ronag) [reply lib/core/client.js:914]: Let me know what tests fail and I can maybe give a hint.

Comment (Ethan-Arrowood) [reply lib/core/client.js:914]: This test fails when I remove the buff and string checks: https://github.com/nodejs/undici/blob/master/test/client-pipelining.js#L453

File: lib/core/client.js (hunk 16 above)
Comment (Ethan-Arrowood) [reply lib/core/client.js:1076]: okay i've gone ahead and modified writeIterable to be as close as I can make it to the stream handling. I'm failing some tests now but I'd appreciate your input on my new changes.

File: lib/core/client.js (hunk 14 above)
Comment (Ethan-Arrowood) [reply lib/core/client.js]: fwiw you asked this to be removed previously so its gone now!

**Comment by Ethan-Arrowood** [comment]
Since its been a while since I made any progress on this I'm going to self-mark this as **stale**. If someone else wants to take a stab at implementing iterators please do so ❤️

**Comment by Linkgoron** [comment]
I've started working on this, unless there are any objections, based on @Ethan-Arrowood original work.

**Comment by ronag** [comment]
This landed in https://github.com/nodejs/undici/pull/848
